crates/ra_hir_ty/src/traits/builtin.rs

```diff
@@ -0,0 +1,161 @@
+//! This module provides the built-in trait implementations, e.g. to make
+//! closures implement `Fn`.
+use hir_def::{expr::Expr, lang_item::LangItemTarget, TraitId, TypeAliasId};
+use hir_expand::name;
+use ra_db::CrateId;
+
+use super::{AssocTyValue, Impl};
+use crate::{db::HirDatabase, ApplicationTy, Substs, TraitRef, Ty, TypeCtor};
+
+pub(super) struct BuiltinImplData {
+    pub num_vars: usize,
+    pub trait_ref: TraitRef,
+    pub where_clauses: Vec<super::GenericPredicate>,
+    pub assoc_ty_values: Vec<AssocTyValue>,
+}
```

Comment:
Hm....

Not related to this PR, but I am slightly worried that we need to do this special casing of closures at all. Ideally, we'd *just* desugar a closure to an a struct with an impl, but we can't express these desugaring with the current hir setup. The problem is that hir generally stores typerefs, and what we need here is an `impl` which refers to a `StructId` directly

Reply:
Ie, I wish we could do something like this: https://github.com/JetBrains/kotlin/blob/fa7bd3dd2e5b230c5ec7cc763bbfe2c4ca374721/compiler/frontend/src/org/jetbrains/kotlin/resolve/DataClassDescriptorResolver.kt#L51-L79 ...

Seems to be a good reason to refactor everything once again :)

Reply:
Closures are not the only case of built-in impls; there are also `Copy` and `Clone` impls for primitives, `Sized` impls, `Unsize`, and `Fn` impls for function types. (That's the reason why I'm extracting this.) For some reason, implementing these using desugaring feels wrong to me, although I guess it'd be possible since we don't need to generate any code for them.

Reply:
Yeah, I mean, deshugaring on the `hir` level, not on the source level. Basically, an ability to synthesize an arbitrary impl on the fly.